[feat](nereids) add rewrite rule :EliminateGroupByKeyByUniform

feiniaofeiafei opened the pull request:
### What problem does this PR solve?
<!--
You need to clearly describe your PR in this part:

1. What problem was fixed (it's best to include specific error reporting information). How it was fixed.
2. Which behaviors were modified. What was the previous behavior, what is it now, why was it modified, and what possible impacts might there be.
3. What features were added. Why this function was added.
4. Which codes were refactored and why this part of the code was refactored.
5. Which functions were optimized and what is the difference before and after the optimization.

The description of the PR needs to enable reviewers to quickly and clearly understand the logic of the code modification.
-->
This PR introduces two main changes:
1. Adds an optional constant value to the uniform attribute in DataTrait. A slot with a constant value that is not null will be considered uniform and not null.
2. Introduces a new transform rule: EliminateGroupByKeyByUniform, which utilizes the newly added part of the uniform attribute. Following is example transformation:
 +--aggregate(group by a,b output a,b,max(c))
 (a is uniform and not null: e.g. a is projection 2 as a in logicalProject)
->
+--aggregate(group by b output b,any_value(a) as a,max(c))



<!--
If there are related issues, please fill in the issue number.
- If you want the issue to be closed after the PR is merged, please use "close #12345". Otherwise, use "ref #12345"
-->

<!--
If this PR is followup a preivous PR, for example, fix the bug that introduced by a related PR,
link the PR here
-->

### Check List (For Author)

- Test <!-- At least one of them must be included. -->

    - [x] Regression test
    - [x] Unit Test
    - [ ] Manual test (add detailed scripts or steps below)
    - [ ] No need to test or manual test. Explain why:
        - [ ] This is a refactor/code format and no logic has been changed.
        - [ ] Previous test can cover this change.
        - [ ] No colde files have been changed.
        - [ ] Other reason <!-- Add your reason?  -->

- Behavior changed:

    - [x] No.
    - [ ] Yes. <!-- Explain the behavior change -->

- Does this need documentation?

    - [x] No.
    - [ ] Yes. <!-- Add document PR link here. eg: https://github.com/apache/doris-website/pull/1214 -->

- Release note

    <!-- bugfix, feat, behavior changed need a release note -->
    <!-- Add one line release note for this PR. -->
    None

### Check List (For Reviewer who merge this PR)

- [ ] Confirm the release note
- [ ] Confirm test cases
- [ ] Confirm document
- [ ] Add branch pick label <!-- Add branch pick label that this PR should merge into -->

feiniaofeiafei committed:
[feat](nereids) add rewrite rule :EliminateGroupByKeyByUniform and PredicateDrivenProjectionSimplification

add license and class describe

group by is empty then not tranform in EliminateGroupByKeyByUniform

remove transform PredicateDrivenProjectionSimplification

feiniaofeiafei commented:
run buildall

morrySnow commented on fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PredicateDrivenProjectionSimplification.java:
```diff
@@ -0,0 +1,73 @@
+import java.util.Map;
+import java.util.Set;
+
+/**ProjectFilterTransform*/
```
class comment should contain what the rule want to do and how

feiniaofeiafei commented:
run buildall

feiniaofeiafei commented:
run buildall

feiniaofeiafei committed:
enhance DataTrait uniform, add const value to uniform description

fix code style

change getAllUniformAndNotNull in data trait builder

change qualify compute uniform

consider a=null predicate and change isUniformAndNotNull logic

use Optional in Uniform Description

if uniform slot can not get const value, use putIfAbsent when add to slotUniformValue

add logical project constant value to uniform description

feiniaofeiafei committed:
add test for rule EliminateGroupByKeyByUniform and change computeUniform in LogicalProject

feiniaofeiafei commented:
run buildall

feiniaofeiafei committed:
add test for rule EliminateGroupByKeyByUniform and change computeUniform in LogicalProject

feiniaofeiafei commented:
run buildall

feiniaofeiafei committed:
fix logical join uniform properties problem

feiniaofeiafei commented:
run buildall

feiniaofeiafei committed:
use CustomRewriter reimplement, and rewrite upper exprid to prevent exprid becoming duplicated

feiniaofeiafei commented:
run buidlall

feiniaofeiafei committed:
fix code style

feiniaofeiafei commented:
run buildall

feiniaofeiafei committed:
not rewrite in cte producer, and change regression mv affected by this rule, and add some cases

feiniaofeiafei commented:
run buildall

feiniaofeiafei committed:
fix code style

feiniaofeiafei commented:
run buildall

feiniaofeiafei committed:
fix compile

feiniaofeiafei commented:
run buildall

feiniaofeiafei committed:
comment out some ut, and add regression comment

feiniaofeiafei commented:
run buildall

feiniaofeiafei commented:
run buildall

feiniaofeiafei commented:
run buildall

feiniaofeiafei committed:
fix ut style

feiniaofeiafei commented:
run buildall

feiniaofeiafei committed:
fix ReplaceRule dead loop

feiniaofeiafei committed:
comment out cases in grace_period, and fix nest replace bug, add nest replace case

feiniaofeiafei commented:
run buildall

feiniaofeiafei committed:
add switch to EliminateGroupByKeyByUniform

feiniaofeiafei committed:
add switch to EliminateGroupByKeyByUniform

feiniaofeiafei committed:
revert aggregate_with_roll_up.groovy

feiniaofeiafei committed:
revert grace_period.groovy

feiniaofeiafei committed:
revert aggregate_without_roll_up.groovy

feiniaofeiafei committed:
revert count_star.groovy

feiniaofeiafei commented:
run p0

feiniaofeiafei committed:
revert feut

feiniaofeiafei committed:
revert materializeViewUtilTest

feiniaofeiafei commented:
run buildall

feiniaofeiafei committed:
fix feut

feiniaofeiafei committed:
fix feut

feiniaofeiafei commented:
run buildall

feiniaofeiafei commented:
run cloud_p0

feiniaofeiafei committed:
remove unrelated change

feiniaofeiafei committed:
remove unrelated change

feiniaofeiafei commented:
run buildall

feiniaofeiafei commented:
run p0

feiniaofeiafei commented:
run cloud_p0

feiniaofeiafei commented:
run external

feiniaofeiafei commented:
run performance

feiniaofeiafei commented:
run buildall

feiniaofeiafei commented:
run cloud_p0

englefly commented on fe/fe-core/src/main/java/org/apache/doris/nereids/properties/DataTrait.java:
```diff
@@ -450,4 +486,104 @@ public NestedSet toImmutable() {
         }
     }
+
+    static class UniformDescription {
```
please explain the usage of UniformDescription by example

xzj7019 commented on fe/fe-core/src/main/java/org/apache/doris/nereids/properties/DataTrait.java:
```diff
@@ -450,4 +486,104 @@ public NestedSet toImmutable() {
+            slotUniformValue = newSlotUniformValue;
+        }
+
+        // The current implementation logic is: if a slot key exists in map slotUniformValue,
```
does it bring the behavior change for the current uniform judgement, in current related rules?

feiniaofeiafei replied:
It will expand slightly. For example, the original isUniformAndNotNull interface, when given the a=1 column a as input, would not return true. After modification, it will return true. The uniform range will not be narrowed, and the meaning of uniform remains unchanged — still indicating a uniform value (which may include additional nulls).

xzj7019 commented on fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/ExprIdRewriter.java:
```diff
@@ -0,0 +1,276 @@
+import java.util.List;
+import java.util.Map;
+
+/**ExprIdReplacer*/
```
pls add some explaination

feiniaofeiafei committed:
add comments

feiniaofeiafei commented:
run buildall

xzj7019 commented on fe/fe-core/src/main/java/org/apache/doris/nereids/properties/DataTrait.java:
```diff
@@ -47,15 +51,15 @@ public class DataTrait {
     public static final DataTrait EMPTY_TRAIT
             = new DataTrait(new NestedSet().toImmutable(),
-                    new NestedSet().toImmutable(), new ImmutableSet.Builder<FdItem>().build(),
+                    new UniformDescription().toImmutable(), new ImmutableSet.Builder<FdItem>().build(),
```
I mean the NestedSet is only used to desc the UniqueSet, may be just recover it to the narrowed down meaning.

feiniaofeiafei commented:
run feut

feiniaofeiafei committed:
remove switch

feiniaofeiafei committed:
remove switch

feiniaofeiafei commented:
run buildall

feiniaofeiafei committed:
fix feut

feiniaofeiafei commented:
run buildall

feiniaofeiafei committed:
rm useless change

xzj7019 commented on fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/ExprIdRewriter.java:
```diff
@@ -0,0 +1,284 @@
+    }
+
+    @Override
+    public List<Rule> buildRules() {
```
confirm the following check_ like rule can detect this kind of unexpected expr id

xzj7019 replied:
if current handling logic is lost and make sure checkAfterRewrite can detect this and report out, to avoid serious execution error, such as core, etc.

xzj7019 commented on fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalJoin.java:
```diff
@@ -535,10 +535,13 @@ public void computeUniform(Builder builder) {
+        // outer join cant have nullable side uniform properties
+        // (e.g. left join may produce null in right side, the uniform value is present and not null
+        // cannot deduce the slot is uniform and not null)
+        if (!joinType.isLeftJoin()) {
```
make sure the null prop lost will not impact current opt rules

feiniaofeiafei committed:
add todo

feiniaofeiafei committed:
fix logicalproject uniform pull up

feiniaofeiafei commented:
run buildall

feiniaofeiafei commented:
run performance

feiniaofeiafei committed:
fix full outer join uniform trait

feiniaofeiafei commented:
run buildall

feiniaofeiafei commented:
run p0

xzj7019 commented on regression-test/suites/nereids_rules_p0/eliminate_gby_key/eliminate_group_by_key_by_uniform.groovy:
```diff
@@ -0,0 +1,221 @@
+    sql "insert into test1 values(1,1),(2,1),(3,1);"
+    sql "create table test2(a int, b int) distributed by hash(a) properties('replication_num'='1');"
+    sql "insert into test2 values(1,105),(2,105);"
+    qt_full_join_uniform_should_not_eliminate_group_by_key "select t2.b,t1.b from test1 t1 full join (select * from test2 where b=105)  t2 on t1.a=t2.a group by t2.b,t1.b order by 1,2;"
```
does the original code have bug for this case?
